packages/components/src/box-control/input-control.tsx

```diff
@@ -0,0 +1,211 @@
+	const allowedSides: Set< keyof BoxControlValue > = new Set(
+		! sides ? ALL_SIDES : []
+	);
+	sides?.forEach( ( allowedSide ) => {
+		if ( allowedSide === 'vertical' ) {
+			allowedSides.add( 'top' );
+			allowedSides.add( 'bottom' );
+		} else if ( allowedSide === 'horizontal' ) {
+			allowedSides.add( 'right' );
+			allowedSides.add( 'left' );
+		} else {
+			allowedSides.add( allowedSide );
+		}
+	} );
+	return allowedSides;
+}
+
+function getSidesToModify(
+	side: BoxControlInputControlProps[ 'side' ],
+	sides: BoxControlInputControlProps[ 'sides' ],
+	isAlt?: boolean
+) {
+	const allowedSides = getAllowedSides( sides );
+
+	let modifiedSides: ( keyof BoxControlValue )[] = [];
+	switch ( side ) {
+		case 'all':
+			modifiedSides = [ 'top', 'bottom', 'left', 'right' ];
+			break;
+		case 'horizontal':
+			modifiedSides = [ 'left', 'right' ];
+			break;
+		case 'vertical':
+			modifiedSides = [ 'top', 'bottom' ];
+			break;
+		default:
+			modifiedSides = [ side ];
+	}
+
+	if ( isAlt ) {
+		switch ( side ) {
+			case 'top':
+				modifiedSides.push( 'bottom' );
+				break;
+			case 'bottom':
+				modifiedSides.push( 'top' );
+				break;
+			case 'left':
+				modifiedSides.push( 'left' );
+				break;
+			case 'right':
+				modifiedSides.push( 'right' );
+				break;
+		}
+	}
+
+	return modifiedSides.filter( ( s ) => allowedSides.has( s ) );
+}
+
+export default function BoxInputControl( {
+	__next40pxDefaultSize,
+	onChange = noop,
+	onFocus = noop,
+	values,
+	selectedUnits,
+	setSelectedUnits,
+	sides,
+	side,
+	...props
+}: BoxControlInputControlProps ) {
+	const defaultValuesToModify = getSidesToModify( side, sides );
+
+	const handleOnFocus = ( event: React.FocusEvent< HTMLInputElement > ) => {
+		onFocus( event, { side } );
+	};
+
+	const handleOnChange = ( nextValues: BoxControlValue ) => {
+		onChange( nextValues );
+	};
+
+	const handleOnValueChange = (
+		next?: string,
+		extra?: { event: React.SyntheticEvent< Element, Event > }
+	) => {
+		const nextValues = { ...values };
+		const isNumeric = next !== undefined && ! isNaN( parseFloat( next ) );
+		const nextValue = isNumeric ? next : undefined;
+		const modifiedSides = getSidesToModify(
+			side,
+			sides,
+			/**
+			 * Supports changing pair sides. For example, holding the ALT key
+			 * when changing the TOP will also update BOTTOM.
+			 */
+			// @ts-expect-error - TODO: event.altKey is only present when the change event was
+			// triggered by a keyboard event. Should this feature be implemented differently so
+			// it also works with drag events?
+			!! extra?.event.altKey
+		);
+
+		modifiedSides.forEach( ( modifiedSide ) => {
+			nextValues[ modifiedSide ] = nextValue;
+		} );
+
+		handleOnChange( nextValues );
+	};
+
+	const handleOnUnitChange = ( next?: string ) => {
+		const newUnits = { ...selectedUnits };
+		defaultValuesToModify.forEach( ( modifiedSide ) => {
+			newUnits[ modifiedSide ] = next;
+		} );
+		setSelectedUnits( newUnits );
+	};
```

Comment:
We might need to make this a bit more complex. It appears like sometimes we want to update them all to the newly selected unit, but sometimes, we want to address them separately. Please, verify in all stories with trunk behavior, but in the meantime, here's an example that seems broken:

* Go to ?path=/story/components-boxcontrol--arbitrary-sides
* Click "Unlick sides"
* Change the unit.

`trunk`:
* Only the modified unit will change

`this branch`:
* Units of all sides will change. 

`trunk`:

Reply:
For me, all the "undefined" values are modified when you change the unit. I would say it's not really clear what should be the right behavior :) but yeah I can update to only change the modified one.

Reply:
I'd aim to keep the same behavior as trunk, and then if we want to modify it, do it separately, instead of being part of a big refactor. How we want it to behave shouldn't really be subject to this PR.

Reply:
Fixed this by simplifying the merging behavior (which doesn't really need to be as complex as what it used to be).